File: sajha_menu/sajha_menu/page/table_order/table_order.py
import frappe
from datetime import datetime
# from jinja2 import Template


# def clever_function():
#     return "Hello"


# template = Template("{{clever_function}}")
# template.globals['clever_function'] = clever_function


@frappe.whitelist(allow_guest=True)
def getTableName():
    names = []
    table_name = frappe.get_list("Tables", fields=['name'])
    for name in table_name:
        names.append(name['name'])
    return names


@frappe.whitelist()
def getBodyHTML():

    table_order = frappe.db.get_list("Tables")
    tableData = []
    for to in table_order:
        tables = frappe.get_doc("Tables", to.name)
        tableData.append(
            {"name": tables.name, "totalOrder": len(tables.orders), "currentOrder": len(tables.current_orders)})
    return frappe.render_template('sajha_menu/page/templates/tables.html', {'tables': tableData})


@frappe.whitelist()
def on_update():
    frappe.publish_realtime("kot_bot_refresh")


@frappe.whitelist()
def applyFilter(company):

    if company == "" or company == None:
        table_order = frappe.db.get_list("Tables")
    else:
        table_order = frappe.db.get_list("Tables", filters=[
            ['company', 'like', company]]
        )

    tableData = []
    for to in table_order:
        tables = frappe.get_doc("Tables", to.name)

        tableData.append(
            {"name": tables.name, "totalOrder": len(tables.orders), "currentOrder": len(tables.current_orders)})
    return frappe.render_template('sajha_menu/page/templates/tables.html', {'tables': tableData})


@frappe.whitelist()
def getCompanyList():
    names = []
    company = frappe.get_list("Company", fields=['name'])
    print(company)
    for name in company:
        names.append(name['name'])
    return names


@frappe.whitelist()
def refreshKot(company):
    if company == "" or company == None:
        table_order = frappe.db.get_list("Tables")
    else:
        table_order = frappe.db.get_list(
            "Tables", filters=[['company', 'like', company]])

    tableData = []
    for to in table_order:
        tables = frappe.get_doc("Tables", to.name)
        tableData.append(
            {"name": tables.name, "totalOrder": len(tables.orders), "currentOrder": len(tables.current_orders)})
    return frappe.render_template('sajha_menu/page/templates/tables.html', {'tables': tableData})

File: sajha_menu/sajha_menu/page/table_order/table_order.js


frappe.pages['table-order'].on_page_load = function (wrapper) {

	new MyPage(wrapper);


}




MyPage = Class.extend({
	init: function (wrapper) {
		this.page = frappe.ui.make_app_page({
			parent: wrapper,
			title: 'Table Orders',
			single_column: true
		});

		frappe.realtime.on('table_refresh', ({ data }) => {
			this.refresh()
		});

		//make page
		this.filters();
		this.make();
	},




	refresh: async function () {

		// window.location.reload()
		let me = $(this);
		let company = this.page.fields_dict['company'].value
		// let fromDate = this.page.fields_dict['fromDate'].value
		// let toDate = this.page.fields_dict['toDate'].value
		// let status = me[0].page.fields_dict['status'].value

		await frappe.call({
			method: "sajha_menu.sajha_menu.page.table_order.table_order.refreshKot",
			args: { "company": company },
			callback: function (r) {
				$("#body-main")[0].innerHTML = frappe.render_template(r.message)
			}
		})
	},


	filters: async function () {
		let me = $(this);

		async function changesTab() {
			let company = me[0].page.fields_dict['company'].value

			await frappe.call({
				method: "sajha_menu.sajha_menu.page.table_order.table_order.applyFilter",
				args: { "company": company },
				callback: function (r) {
					$("#body-main")[0].innerHTML = frappe.render_template(r.message)
				}
			})
			// push dom element to page
		}



		await frappe.call({
			method: "sajha_menu.sajha_menu.page.table_order.table_order.getCompanyList",
			callback: function (r) {
				companyName = r.message
			}
		})

		this.page.add_field({
			label: 'Company',
			fieldtype: 'Link',
			fieldname: 'company',
			options: "Company",
			change() {
				changesTab();
			}
		});

	},

	make: async function () {
		// grab the class 
		let me = $(this);


		// body content
		let body = await frappe.call({
			method: "sajha_menu.sajha_menu.page.table_order.table_order.getBodyHTML",
			callback: function (r) {
			}
		})
		// push dom element to page
		$(frappe.render_template(body.message, this)).appendTo(this.page.main)


		// execute methods
		// tableName();
	}

	// end of class
})



